src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -496,6 +502,18 @@ where
                         }
                     }
                     AsyncEvent::Progress(frontier) => {
+                        // We need to be *very* careful to check the shutdown
+                        // token here. A shutdown will cause a progress message
+                        // that advances to the empty frontier, and depending on
+                        // when the shutdown occurs we may have partial data
+                        // staged on S3. We *must not* call `finish` and must
+                        // instead abandon the upload to avoid committing that
+                        // partial data.
+                        if shutdown_token.strong_count() == 0 {
+                            // Wedge until the operator is dropped.
+                            future::pending::<()>().await;
```

Comment:
What is the mechanism that will drop the iterator? I think it could be the button `AsyncOperatorBuilder::build` returns, but we are not using that for this operator. Doesn't that mean we will keep this future alive forever?

Reply:
Responded on Slack, but whoops, thanks! Updated.